Terminal fixes

ziz57 opened the pull request:
Main effect is to make terminal usably fast with TUI programs which make extensive use of escape codes, which previously it wasn't (for me).

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/12369)
<!-- Reviewable:end -->

poire-z commented:
Pinging @zwim for review.

Frenzie commented:
Looks interesting, waiting for zwim. :-)

> TUI programs which make extensive use of escape codes

Could you name any examples? I didn't notice any slowness on my H2O in various programs I tried such as Elinks and Vim.

ziz57 commented:
Mostly I was having problems with screen. It uses a lot of ^[K, so getting rid of the extra initTextBox there was the key change which fixed this for me. The other optimisations and fixes are things I tried/noticed while debugging this, and seemed worth doing.

ziz57 commented:
>I have mainly tested the emulator with 'bash' 'vi' 'nano' 'shfm' 
>'watch' and the common bash commands.
>One of them uses the ^[= and ^[> sequences.

Hmm. Could it be that one of them was compiled for a different terminal? 
This use of these sequences doesn't seem to part of VT100 or ANSI.

>I would prefer
>`UIManager:nextTick(self.enter_callback)`

Done.

ziz57 commented:
@zwim I added one more fix, to the wrapping code; please review.

I'm also wondering what's going on with reverseLineFeed. It doesn't seem 
to implement the vt52 specified behaviour. Is this for some sort of bug 
compatibility, or should it be fixed?

Have you thought about better handling of double-width characters?

poire-z commented:
> I will retrire from this Project (in a friendly way).

It's the fourth time you tell us that. So, see you soon :)
(You know you can't leave us, even if you want to.)

ziz57 commented:
* Saturday, 2024-08-24 at 11:29 -0700 - zwim ***@***.***>:

>On Line 16 we have backspace defined (though is is Not an esc 
>sequence).
>You might want to use that (or even better defined it right and use it) 

Done.

ziz57 commented:
* Saturday, 2024-08-24 at 11:44 -0700 - zwim ***@***.***>:

>@ziz57 You are Doing a good Job. Do you want to inherit this Code. 
>I would Support you.

I don't know about inheriting it, but I guess I know it well enough now 
that I can review changes. I'm also tempted to try to rework it to 
handle wide characters, but we'll see.

poire-z commented:
I guess nobody else will be able to review this PR :)
So @ziz57 when you're done, tell us so we can merge it.
You may also want to help reviewing #12384.

ziz57 commented:
>I guess nobody else will be able to review this PR :)
>So @ziz57 when you're done, tell us so we can merge it.

This PR is finished and can be merged.

>You may also want to help reviewing #12384.

I had a quick look, but I don't know the context with Android and have 
nothing useful to add.

Frenzie commented on plugins/terminal.koplugin/main.lua:
```diff
@@ -393,6 +393,7 @@ function Terminal:generateInputDialog()
                         end
 
                         UIManager:close(self.input_dialog)
+                        UIManager:unschedule(Terminal.refresh)
```
I think the unschedule first and then close still stands.

ziz57 commented:
>I think the unschedule first and then close still stands.

Thanks; fixed.

I also added a @fixme to ReverseLineFeed.

Now it should really be ready to merge.
